pv211_utils/evaluation_metrics.py

```diff
@@ -0,0 +1,405 @@
+        current_rank += 1
+
+    if num_relevant > 0:
+        recall = num_relevant_topk / num_relevant
+    else:
+        recall = 1
+
+    return recall
+
+
+def _calc_precision(system: IRSystemBase, judgements: Set, k: int,
+                    query: QueryBase) -> float:
+    num_relevant = 0
+    precision = 0.0
+    current_rank = 1
+
+    for document in system.search(query):
+        if current_rank > k:
+            break
+        if (query.query_id, document.document_id) in judgements:
+            num_relevant += 1
+        current_rank += 1
+
+    precision = num_relevant / k
+
+    return precision
+
+
+def _calc_average_precision(system: IRSystemBase, judgements: Set, k: int,
+                            query: QueryBase) -> float:
+    num_relevant = 0
+    average_precision = 0.0
+    current_rank = 1
+
+    for document in system.search(query):
+        if current_rank > k:
+            break
+        if (query.query_id, document.document_id) in judgements:
+            num_relevant += 1
+            average_precision += num_relevant / current_rank
+        current_rank += 1
+
+    average_precision /= num_relevant if num_relevant > 0 else 1
+
+    return average_precision
+
+
+class calc_map(abc.ABC):
+    system = None
+    judgements = None
+    k = None
+    _CURRENT_INSTANCE = None
+
+    @classmethod
+    def _calc_average_precision(csl, query: QueryBase) -> float:
+        num_relevant = 0
+        average_precision = 0.0
+        current_rank = 1
+
+        for document in csl._CURRENT_INSTANCE.system.search(query):
+            if current_rank > csl._CURRENT_INSTANCE.k:
+                break
+            if (query.query_id, document.document_id) in csl._CURRENT_INSTANCE.judgements:
+                num_relevant += 1
+                average_precision += num_relevant / current_rank
+            current_rank += 1
+
+        average_precision /= num_relevant if num_relevant > 0 else 1
+
+        return average_precision
+
+    def mean_average_precision(self, system: IRSystemBase, queries: OrderedDict,
+                               judgements: Set[JudgementBase],
+                               k: int, num_processes: int) -> float:
+        """Evaluate system for given queries and judgements with mean average precision
+        metric. Where first k documents will be used in evaluation.
+
+        Arguments
+        ---------
+        system : IRSystemBase
+            System to be evaluated.
+        queries : OrderedDict
+            Queries to be searched.
+        judgements : Set[JudgementBase]
+            Judgements.
+        k : int
+            Parameter defining evaluation depth.
+        num_processes : int
+            Parallelization parameter defining number of processes to be used to run the evaluation.
+
+        Returns
+        -------
+        float
+            Mean average precision score from interval [0, 1].
+        """
+        map_score = 0.0
+
+        self.system = system
```

Comment:
Pickling a system sounds easy to break or, worse, to slow it down to a point where it would be faster to evaluate with one worker when the system references a large object. Let's save the current system to `self.__class__._CURRENT_SYSTEM` to skip the pickling? If we are keeping `mean_average_precision`, make `_CURRENT_SYSTEM` a global?

Reply:
When this has been fixed, let's add a modest `num_workers` default to `notebooks/*.ipynb` to speed things up for students. Too brittle at the moment, so keeping the `num_workers=1` for the moment.